Source/Private/ASTextNode2.mm

```diff
@@ -0,0 +1,1122 @@
+  }
+  
+  // Tell the display node superclasses that the cached layout is incorrect now
+  [self setNeedsLayout];
+  
+  // Force display to create renderer with new size and redisplay with new string
+  [self setNeedsDisplay];
+  
+  
+  // Accessiblity
+  self.accessibilityLabel = attributedText.string;
+  self.isAccessibilityElement = (length != 0); // We're an accessibility element by default if there is a string.
+}
+
+#pragma mark - Text Layout
+
+- (void)setExclusionPaths:(NSArray *)exclusionPaths
+{
+  _textContainer.exclusionPaths = exclusionPaths;
+  
+  [self setNeedsLayout];
+  [self setNeedsDisplay];
+}
+
+- (NSArray *)exclusionPaths
+{
+  return _textContainer.exclusionPaths;
+}
+
+- (void)prepareAttributedStringForDrawing:(NSMutableAttributedString *)attributedString
+{
+  ASDN::MutexLocker lock(__instanceLock__);
+ 
+  // Apply paragraph style if needed
+  [attributedString enumerateAttribute:NSParagraphStyleAttributeName inRange:NSMakeRange(0, attributedString.length) options:kNilOptions usingBlock:^(NSParagraphStyle *style, NSRange range, BOOL * _Nonnull stop) {
+    if (style == nil || style.lineBreakMode == _truncationMode) {
+      return;
+    }
+    
+    NSMutableParagraphStyle *paragraphStyle = [style mutableCopy] ?: [[NSMutableParagraphStyle alloc] init];
+    paragraphStyle.lineBreakMode = _truncationMode;
+    [attributedString addAttribute:NSParagraphStyleAttributeName value:paragraphStyle range:range];
+  }];
+  
+  // Apply background color if needed
+  UIColor *backgroundColor = self.backgroundColor;
+  if (CGColorGetAlpha(backgroundColor.CGColor) > 0) {
+    [attributedString addAttribute:NSBackgroundColorAttributeName value:backgroundColor range:NSMakeRange(0, attributedString.length)];
+  }
+  
+  // Apply shadow if needed
+  if (_shadowOpacity > 0 && (_shadowRadius != 0 || !CGSizeEqualToSize(_shadowOffset, CGSizeZero)) && CGColorGetAlpha(_shadowColor) > 0) {
+    NSShadow *shadow = [[NSShadow alloc] init];
+    if (_shadowOpacity != 1) {
+      shadow.shadowColor = [UIColor colorWithCGColor:CGColorCreateCopyWithAlpha(_shadowColor, _shadowOpacity * CGColorGetAlpha(_shadowColor))];
+    } else {
+      shadow.shadowColor = [UIColor colorWithCGColor:_shadowColor];
+    }
+    shadow.shadowOffset = _shadowOffset;
+    shadow.shadowBlurRadius = _shadowRadius;
+    [attributedString addAttribute:NSShadowAttributeName value:shadow range:NSMakeRange(0, attributedString.length)];
+  }
+}
+
+#pragma mark - Drawing
+
+- (NSObject *)drawParametersForAsyncLayer:(_ASDisplayLayer *)layer
+{
+  [self _ensureTruncationText];
+  ASTextContainer *copiedContainer = [_textContainer copy];
+  copiedContainer.size = self.bounds.size;
+  NSMutableAttributedString *mutableText = [self.attributedText mutableCopy] ?: [[NSMutableAttributedString alloc] init];
+  [self prepareAttributedStringForDrawing:mutableText];
```

Comment:
It's not super expensive, and we currently only do it on main when gathering drawParameters. To do it off-main in that case would be tough and probably not worth it, since it needs to grab state from the node `shadowColor` and potentially a lot more.